### DISCUSSION TITLE
Filling corner ghost cells

### DISCUSSION THREAD

**Discussion by amneetb** [IBAMR/IBAMR#1765] [Q&A]
I am having issues with filling ghost cells of side-centered variables at the corners. I am observing at after using `HierarchyGhostCellInterpolation` the ghost cell values at corners like southwest still remain unfilled as `NaNs`. This is what I am doing 

```
// Fill patch data of  side-centered U1_real_idx and U1_imag_idx

for (int ln = 0; ln <= patch_hierarchy->getFinestLevelNumber(); ++ln)
        {
            Pointer<PatchLevel<NDIM> > level = patch_hierarchy->getPatchLevel(ln);
            level->allocatePatchData(U1_real_idx);
            level->allocatePatchData(U1_imag_idx);
            for (PatchLevel<NDIM>::Iterator p(level); p; p++)
            {
                Pointer<Patch<NDIM> > patch = level->getPatch(p());
                Pointer<SideData<NDIM, double> > U1_real_data = patch->getPatchData(U1_real_idx);
                Pointer<SideData<NDIM, double> > U1_imag_data = patch->getPatchData(U1_imag_idx);
                const Box<NDIM>& ghost_box = U1_real_data->getGhostBox();
                const Box<NDIM>& patch_box = patch->getBox();
                const hier::Index<NDIM>& patch_lower = patch_box.lower();
                Pointer<CartesianPatchGeometry<NDIM> > pgeom = patch->getPatchGeometry();
                const double* const patch_x_lower = pgeom->getXLower();
                const double* const dx = pgeom->getDx();

                std::array<double, NDIM> posn;
                for (unsigned int axis = 0; axis < NDIM; ++axis)
                {
                    for (SideIterator<NDIM> ic(patch_box, axis); ic; ic++)
                    {
                        const SideIndex<NDIM>& i = ic();
                        for (unsigned int d = 0; d < NDIM; ++d)
                        {
                            if (d == axis)
                            {
                                posn[d] = patch_x_lower[d] + dx[d] * (static_cast<double>(i(d) - patch_lower(d)));
                            }
                            else
                            {
                                posn[d] = patch_x_lower[d] + dx[d] * (static_cast<double>(i(d) - patch_lower(d)) + 0.5);
                            }
                        }
                        const double& X0 = posn[0];
                        const double& X1 = posn[1];

                        if (axis == 0)
                        {
                            (*U1_real_data)(i) = std::pow(X0,3) + std::pow(X1,3);  
                            (*U1_imag_data)(i) = std::pow(X0,2) + std::pow(X1,2); 
                        }
                        
                        if (axis == 1)
                        {
                            (*U1_real_data)(i) = std::pow(X0,2) + std::pow(X1,2);
                            (*U1_imag_data)(i) = std::pow(X0,3) + std::pow(X1,3);
                        }
                    }
                }
            }
        }


        // Fill the ghost cells of U1_real and U1_imag
        using InterpolationTransactionComponent = HierarchyGhostCellInterpolation::InterpolationTransactionComponent;
        std::vector<InterpolationTransactionComponent> comp_transactions(2);
        comp_transactions[0] = InterpolationTransactionComponent(U1_real_idx,
                                                                 "CONSERVATIVE_LINEAR_REFINE",
                                                                 true,
                                                                 "CONSERVATIVE_COARSEN",
                                                                 "QUADRATIC",
                                                                 false,
                                                                 U1_bc_coefs[0]);
        comp_transactions[1] = InterpolationTransactionComponent(U1_imag_idx,
                                                                 "CONSERVATIVE_LINEAR_REFINE",
                                                                 true,
                                                                 "CONSERVATIVE_COARSEN",
                                                                 "QUADRATIC",
                                                                 false,
                                                                 U1_bc_coefs[1]);
        HierarchyGhostCellInterpolation comp_fill_op;
        comp_fill_op.initializeOperatorState(comp_transactions, patch_hierarchy);
        comp_fill_op.fillData(/*time*/0.0);  

```  

Is there any option I am missing to fill out the ghost values at corners?

**Comment by boyceg** [comment]
That should do it, but I don't think we have any test coverage for this case. This could be a bug in either IBAMR or SAMRAI. I would suggest making sure that the fill pattern really is `nullptr` inside `HierarchyGhostCellInterpolation`.

**Comment by amneetb** [comment]
I tried adding fill pattern as a nullptr explicitly, but it still does not fill it properly. 

I am attaching a bare bone example which shows the problem. Run it with 2 processors so that the domain is split in half and patches have corner ghost cells. 

Can you please help fix this bug? It is crucial for my acoustic solvers as the variable coefficient stencils are wide and involve corner DoFs.  

[test_case.zip](https://github.com/user-attachments/files/17004220/test_case.zip)

**Comment by boyceg** [comment]
Does a basic refine algorithm/schedule reproduce? Does it matter if the filling is single level (without a refine operator) vs multi-level?

**Comment by amneetb** [comment]
I am testing only for a single level. The code is written for multi levels.

--Amneet

**Comment by boyceg** [comment]
I would suggest testing what happens with a standalone refine schedule in a simple test code. If you cannot get that to work, then it will be necessary to figure out what is going wrong inside SAMRAI.

**Comment by boyceg** [comment]
Although --- it is surprising that you are having this problem, since we use this kind of filling for L/E interaction for `IBMethod`, `IBFEMethod`, `IIMethod`, etc. I would check out how ghost cell filling is set up in `LEInteractor` and `FEDataManager`.

**Comment by amneetb** [comment]
I took a look, and what I am doing appears to be consistent. I am not sure what is going on, but there is a bug somewhere in the library.

**Reply by boyceg** [reply]
> I took a look, and what I am doing appears to be consistent. I am not sure what is going on, but there is a bug somewhere in the library.

Could you check to see if the problem is not happening in there, or if it is just being hidden because the ghost cell width is larger?

**Reply by amneetb** [reply]
I reduced the ghost cell width from 2 to 1 and it appears to re-solve the problem. In my applications, however, I need ghost cell width to be 2 or more.

**Reply by boyceg** [reply]
I would suggest to printing out side ghost cell values filled by in `FEDataManager` or `LDataManager`. Those classes commonly fill larger ghost cell regions and, as far as I know, there is no issue with corners -- otherwise IB-type methods would not work with larger regularized delta functions.

**Reply by amneetb** [reply]
Yes, but this exercise does not solve the problem at hand.

Something is either not set correctly or there is some subtle bug in the library that does not show up often in IB applications.

One thing I can think of that this ghost filling bug is coming near the domain boundaries. Not sure if IB interpolation uses wider stencil near the physical boundaries or not.

**Reply by boyceg** [reply]
Oh, is this something that only occurs with corners of the computational domain?

If so, the code to look at is the "codimension 2" boundary filling routines (e.g., `CartSideRobinPhysBdryOp::fillGhostCellValuesCodim2()`).

Because there are not any boundary conditions for the corners, those routines are intended to extrapolate from the "codimension 1" ghost cell values that were set according to physical boundary conditions.

**Reply by amneetb** [reply]
No this happens even at the patch corner touching the physical boundary right in the middle of the domain; see below

```
-----------------------------
|             |              |
|             |              |
|             |              |
------------------------------
              ^
              |
  (problematic corner)

```

**Comment by amneetb** [comment]
I added refine schedules to fill ghost data as well. The problem still persists. Maybe you can give it a run.

**Comment by amneetb** [comment]
I am convinced that there is a bug somewhere in IBAMR that does not consistently account for ghost cell width > 1 when filling codim 2 boundary conditions. 

The test I sent earlier is a good (and a simple) test to detect this.

**Reply by boyceg** [reply]
Understood. I can take a look at some point, but if you are in a hurry, I would strongly recommend checkout out what happens in either `LDataManager` or in `FEDataManager` when performing IB interpolation for kernels with large stencils. There are standalone interpolation tests for IB interpolation in the test suite that will make it easy to see what ghost cell values are being filled (or not) for simple fields (like constant).

One thing to check out is the way that the "scratch" context is being set. TBH, I think that I remember seeing some differences depending on whether the scratch data is distinct from the destination data. A thing to look at is how the choices of the source, scratch, and destination patch data indices impact the results (if at all).

**Comment by amneetb** [comment]
I think I have narrowed it down to the lack of width in one of the boundary fill box in `CartExtrapPhysBdryOp:

https://github.com/IBAMR/IBAMR/blob/master/ibtk/src/boundary/physical_boundary/CartExtrapPhysBdryOp.cpp#L211
https://github.com/IBAMR/IBAMR/blob/master/ibtk/src/boundary/physical_boundary/CartExtrapPhysBdryOp.cpp#L653

```
patch_box = [(0,0),(63,127)]
bdry_fill_box = [(0,-2),(64,-1)]
bdry_fill_box * ghost_box = [(0,-2),(64,-1)]
```

The value of `ghost_width_to_fill` is (2,2). However, the boundary fill box size is only one more in the x-direction, past the patch box. This is the cause of the problem for the problematic ghost cell. The cell index (65,-1) is not covered in the set of boundary boxes generated here:

```
patch_box = [(0,0),(63,127)]
bdry_fill_box = [(-2,-2),(-1,-1)]
bdry_fill_box * ghost_box = [(-2,-2),(-1,-1)]
patch_box = [(0,0),(63,127)]
bdry_fill_box = [(-2,128),(-1,129)]
bdry_fill_box * ghost_box = [(-2,128),(-1,129)]
patch_box = [(0,0),(63,127)]
bdry_fill_box = [(-2,0),(-1,127)]
bdry_fill_box * ghost_box = [(-2,0),(-1,127)]
patch_box = [(0,0),(63,127)]
bdry_fill_box = [(0,-2),(64,-1)]
bdry_fill_box * ghost_box = [(0,-2),(64,-1)]
patch_box = [(0,0),(63,127)]
bdry_fill_box = [(0,128),(64,129)]
bdry_fill_box * ghost_box = [(0,128),(64,129)]
```

There is no index 65 in the x-direction. If ghost cell width is (2,2), the bdry_fill_box should have an index 65 in the x-direction. 


I have set BCs to be `nullptr` so there is no role of `CartSideRobinPhysBdryOp` in generating NaNs.

**Reply by boyceg** [reply]
This would only impact physical boundaries, though -- is this only an issue at physical boundaries, or is it happening at all interior corners as well?

**Reply by amneetb** [reply]
Only at the corners abutting physical boundaries.

**Reply by boyceg** [reply]
Ah. OK, then yeah, the places to look are `CartExtrapPhysBdryOp` and `CartSideRobinPhysBdryOp`. It will be easiest to debug if you just apply them to data that is constant non-zero in the interior.

**Reply by amneetb** [reply]
I think the bug is in generating boundary boxes. We are using patch geometry object to do this 
https://github.com/IBAMR/IBAMR/blob/master/ibtk/src/boundary/physical_boundary/CartExtrapPhysBdryOp.cpp#L248

Perhaps we need to patch up SAMRAI in order to make it respect the ghost cell width size of boundary boxes.

**Reply by boyceg** [reply]
`setPhysicalBoundaryConditions_side()` needs to interpret the boundary box correctly. The basic idea is that the boundary box should cover the cell-centered region that needs to be filled. If that isn't happening, then that's the first thing that is going wrong. Once the cell-centered box is generated, the different specialized routines need to convert it correctly into edge/node/side/etc index spaces.

**Comment by amneetb** [comment]
This definition looks suspicious in SAMARAI. The boundary fill box is grown by (gcw - 1) instead of gcw...

```

template<int DIM> Box<DIM>
PatchGeometry<DIM>::getBoundaryFillBox(const BoundaryBox<DIM>& bbox,
                                        const Box<DIM>& patch_box,
                                        const IntVector<DIM>& gcw) const
{
#ifdef DEBUG_CHECK_ASSERTIONS
   for (int i = 0; i < DIM; i++) {
      TBOX_ASSERT( gcw(i) >= 0 );
   }
#endif

   Box<DIM> tmp_box(patch_box);
   tmp_box.grow(gcw);
   Box<DIM> fill_box(bbox.getBox() * tmp_box);

   int bdry_type = bbox.getBoundaryType();
   int location_index = bbox.getLocationIndex();

   // Get the singleton class lookup table 
   const BoundaryLookupTable<DIM>* blut;
   blut = BoundaryLookupTable<DIM>::getLookupTable();

#ifdef DEBUG_CHECK_ASSERTIONS
   const tbox::Array<int> &location_index_max = blut->getMaxLocationIndices();
   TBOX_ASSERT(bdry_type > 0);
   TBOX_ASSERT(bdry_type <= DIM);
   TBOX_ASSERT(location_index >= 0);
#endif

   if (!fill_box.empty()) {

      // Loop over codimension (a.k.a. boundary type)
      for (int codim = 1; codim <= DIM; codim++) {

	 // When we get a match on the boundary type
         if (bdry_type == codim) {

#ifdef DEBUG_CHECK_ASSERTIONS
            TBOX_ASSERT(location_index < location_index_max[codim-1]);
#endif
            // Get the directions involved in this boundary type from the 
	    // lookup table.
	    const tbox::Array<int> &dir =
               blut->getDirections(location_index, codim);

	    // For each direction, identify this as an upper or lower boundary.
            for (int i = 0; i < codim; i++) {
	       if (blut->isUpper(location_index, codim, i)) {
                  fill_box.growUpper(dir[i], gcw(dir[i])-1);
	       } else {
                  fill_box.growLower(dir[i], gcw(dir[i])-1);
	       }
            }

	    // We've found boundary type, so break out of the loop.
            break;
         }
      }
   }

   return (fill_box);
}

```

**Reply by boyceg** [reply]
I'm not sure --- if these boundary boxes are not computed in the right way, then it seems like nothing would work at boundaries.

The problem might be that we need to convert this into the right kind of side box (and are not doing it correctly).

**Reply by amneetb** [reply]
I don't think that is the issue. The issue is excluding a corner from the list of boundary boxes. See the corner below. It is excluded from both codim 1 and codim 2 boundaries.  

```
-----------------------------
|             |              |
|   P1        |    P2        |
|             |              |
------------------------------
           |   |
           |___|       ----> This is not included as a codim 2 boundary for either patch 1 (P1) or 2 (P2)
             ^
             |
  (problematic corner)
```

**Reply by boyceg** [reply]
It should be a codim 1 boundary, not a codim 2 boundary. Codim 2 boundary are just the actual corners of the domain. In principle there are physical boundary conditions that should be used there. (We must not switch to extrapolation at the problematic corner, otherwise we can get different answers depending on where exactly the patch boundary is.)

**Reply by amneetb** [reply]
I agree. The problem is that the codim 1 boundary box size is not grown properly here. It should have increased by 2, but it did only by 1 for some reason.

**Reply by boyceg** [reply]
I think that there are some creaky old routines in IBAMR / IBTK that "extend" boundary filling regions. I wonder if those need to be used here.

In some sense, the right approach here might be to work backwards. I think that the codimension 1 region should be the whole ghost cell region that touches the physical boundary, excluding the regions that are codim > 1.

It is a little nasty, but I guess what could be done is to start with SAMRAI's codim 1 region, extend it "away" from the patch interior in all directions, trim with the ghost box, and then trim away codim > 1 regions.

(Also, those codim > 1 regions might similarly need to be extended?)

I am not going to be able to write any code this week, but I can try to take a look at something that implements this if you have time to do it.
